pkg/hubble/observer/local_observer.go

```diff
@@ -681,3 +710,65 @@ func newRingReader(ring *container.Ring, req genericRequest, whitelist, blacklis
 	}
 	return container.NewRingReader(ring, idx), nil
 }
+
+type namespaceRecord struct {
+	namespace *observerpb.Namespace
+	added     time.Time
+}
+
+type namespaceManager struct {
+	mu         lock.RWMutex
+	namespaces map[string]namespaceRecord
+}
+
+func NewNamespaceManager() *namespaceManager {
+	return &namespaceManager{
+		namespaces: make(map[string]namespaceRecord),
+	}
+}
+
+func (m *namespaceManager) Run(ctx context.Context) {
+	ticker := time.NewTicker(5 * time.Minute)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case t := <-ticker.C:
+			// periodically remove any namespaces which haven't been seen in flows
+			// for the last hour
+			m.mu.Lock()
+			for key, record := range m.namespaces {
+				if record.added.Add(time.Hour).After(t) {
+					delete(m.namespaces, key)
+				}
+			}
+			m.mu.Unlock()
```

Comment:
I think using the ticker's time is cute, but a package-level `var NowFunc = time.Now` and `now := NowFunc()` before iterating would be effectively the same while easier to test.

I'm sure we'll be asked to make it configurable, but in the meantime let's add a constant as well for this 1h "timeout".

Finally, extracting this into a `namespaceManager` func would help wrt testing

```suggestion
			m.Cleanup(NowFunc())
```

````go
// package-level
var NowFunc = time.Now
func (m *namespaceManager) Cleanup(now time.Time) {
	m.mu.Lock()
	defer m.mu.Unlock()
	for key, record := range m.namespaces {
		if record.added.Add(time.Hour).After(now) {
			delete(m.namespaces, key)
		}
	}
}

Reply:
Yeah I was considering how to do the `Now()` check. I wish we had a `clock` interface, because I really dislike the globals, but I'll take that approach for now.